src/map/script.c

```diff
@@ -18318,33 +18318,66 @@ static BUILDIN(swap)
 
 // [zBuffer] List of dynamic var commands --->
 
+/**
+ * Sets the value of a variable or an array.
+ * The variable's name can be passed as a string and thus can dynamically be constructed.
+ *
+ * [Kenpachi]
+ * Passing a reference to a variable is supported, too.
+ * Useful when using getvariableofnpc(), getvariableofpc(), getd() or getarg().
+ *
+ * @code{.herc}
+ *	setd("<variable name>", <value>);
+ *	setd(<reference to variable>, <value>);
+ * @endcode
+ *
+ */
 static BUILDIN(setd)
 {
-	struct map_session_data *sd = NULL;
+	int idx = 0;
+	int length = 0;
 	char varname[100];
 	const char *buffer;
-	int elem;
-	buffer = script_getstr(st, 2);
+	const void *value;
+	struct map_session_data *sd = NULL;
+	struct script_data *data = script_getdata(st, 2);
+	struct reg_db *ref = reference_getref(data);
 
-	if(sscanf(buffer, "%99[^[][%d]", varname, &elem) < 2)
-		elem = 0;
+	if (data_isreference(data)) {
+		buffer = reference_getname(data);
+		idx = reference_getindex(data);
 
-	if( not_server_variable(*varname) )
-	{
-		sd = script->rid2sd(st);
-		if( sd == NULL )
-		{
-			ShowError("script:setd: no player attached for player variable '%s'\n", buffer);
-			return true;
-		}
+		if (not_server_variable(*buffer))
+			ref = NULL;
+	} else {
+		buffer = script_getstr(st, 2);
```

Comment:
this look like allow for example ``setd(10, 20)``. probably here need check for is ``st,2`` really string?

Reply:
Changed and tested.